uc-crux-llvm/src/UCCrux/LLVM/FullType/Type.hs

```diff
@@ -0,0 +1,279 @@
+    MapToBaseType sym arch xs Ctx.::> ToBaseType sym arch x
+
+-- | The type of annotated What4 values that correspond to each 'FullType'
+type family ToBaseType (sym :: Type) (arch :: LLVMArch) (ft :: FullType m) :: CrucibleTypes.BaseType where
+  ToBaseType sym arch ('FTInt n) = CrucibleTypes.BaseBVType n
+  ToBaseType sym arch ('FTPtr _ft) = CrucibleTypes.BaseIntegerType
+  ToBaseType sym arch ('FTFloat flt) = W4IFP.SymInterpretedFloatType sym flt
+  ToBaseType sym arch ('FTStruct ctx) =
+    CrucibleTypes.BaseStructType (MapToBaseType sym arch ctx)
+
+data FullTypeRepr (m :: Type) (ft :: FullType m) where
+  FTIntRepr ::
+    (1 <= w) =>
+    !(NatRepr w) ->
+    FullTypeRepr m ('FTInt w)
+  FTPtrRepr ::
+    PartTypeRepr m ft ->
+    FullTypeRepr m ('FTPtr ft)
+  FTFloatRepr ::
+    !(CrucibleTypes.FloatInfoRepr flt) ->
+    FullTypeRepr m ('FTFloat flt)
+  FTArrayRepr ::
+    (1 <= n) =>
+    !(NatRepr n) ->
+    FullTypeRepr m ft ->
+    FullTypeRepr m ('FTArray n ft)
+  FTStructRepr ::
+    MemType.StructInfo ->
+    Ctx.Assignment (FullTypeRepr m) fields ->
+    FullTypeRepr m ('FTStruct fields)
+  FTVoidFuncPtrRepr ::
+    Ctx.Assignment (FullTypeRepr m) args ->
+    FullTypeRepr m ('FTFuncPtr 'Nothing args)
+  FTNonVoidFuncPtrRepr ::
+    FullTypeRepr m ret ->
+    Ctx.Assignment (FullTypeRepr m) args ->
+    FullTypeRepr m ('FTFuncPtr ('Just ret) args)
+  -- TODO(lb): This could have a symbol repr for the name
+  FTOpaquePtrRepr :: L.Ident -> FullTypeRepr m 'FTOpaquePtr
+
+-- | This functions similarly to 'MemType.SymType'
+data PartTypeRepr (m :: Type) (ft :: FullType m) where
+  PTFullRepr :: FullTypeRepr m ft -> PartTypeRepr m ft
+  -- The Const is so that we can get type variables in scope in the TestEquality
+  -- instance, see below.
+  PTAliasRepr :: Const L.Ident ft -> PartTypeRepr m ft
+
+data IsPtrRepr m ft = forall ft'. IsPtrRepr (ft :~: 'FTPtr ft')
+
+isPtrRepr :: forall m ft. FullTypeRepr m ft -> Maybe (IsPtrRepr m ft)
+isPtrRepr =
+  \case
+    FTPtrRepr _ -> Just (IsPtrRepr Refl)
+    _ -> Nothing
+
+$(return [])
+
+-- | We assume (via unsafeCoerce) that types with the same L.Ident are the same.
+-- This is validated by the existential used in @makeModuleTypes@.
+instance TestEquality (PartTypeRepr m) where
+  testEquality =
+    $( U.structuralTypeEquality
+         [t|PartTypeRepr|]
+         ( let appAny con = U.TypeApp con U.AnyType
+            in [ ( appAny (appAny (U.ConType [t|FullTypeRepr|])),
+                   [|testEquality|]
+                 ),
+                 ( appAny (appAny (U.ConType [t|Ctx.Assignment|])),
+                   [|testEquality|]
+                 ),
+                 ( appAny (U.TypeApp (U.ConType [t|Const|]) (U.ConType [t|L.Ident|])),
+                   [|
+                     \(Const ident1 :: Const L.Ident ft1) (Const ident2 :: Const L.Ident ft2) ->
+                       if ident1 == ident2 then Just (unsafeCoerce Refl :: ft1 :~: ft2) else Nothing
```

Comment:
This unsafe coerce makes me very nervous.  The `PartTypeRepr` constructors are exposed, and nothing prevents a user from improperly using them.  My usual standard for unsafe coerce magics is that they need to be justified by invariants that are encapsulated in a single module and protected as well as possible by restricting export lists/smart constructors/etc.

Reply:
I'll move all the relevant code into the same module and restrict the exports.

Reply:
The relevant constructors are no longer exported!